node/src/chain/fsm.rs

```diff
@@ -160,45 +167,67 @@ impl<N: Network, DB: database::DB, VM: vm::VMExecution> SimpleFSM<N, DB, VM> {
-                    // Transition from InSync to OutOfSync state
-                    curr.on_exiting().await;
+        if let Some(blk) = self.get_block_with_cert(blk).as_ref() {
+            match &mut self.curr {
+                State::InSync(ref mut curr) => {
+                    if let Some((b, peer_addr)) =
+                        curr.on_block_event(blk, metadata).await?
+                    {
+                        // Transition from InSync to OutOfSync state
+                        curr.on_exiting().await;
 
-                    // Enter new state
-                    let mut next = OutOfSyncImpl::new(
-                        self.acc.clone(),
-                        self.network.clone(),
-                    );
-                    next.on_entering(&b, peer_addr).await;
-                    self.curr = State::OutOfSync(next);
+                        // Enter new state
+                        let mut next = OutOfSyncImpl::new(
+                            self.acc.clone(),
+                            self.network.clone(),
+                        );
+                        next.on_entering(&b, peer_addr).await;
+                        self.curr = State::OutOfSync(next);
+                    }
                 }
-            }
-            State::OutOfSync(ref mut curr) => {
-                if curr.on_block_event(blk, metadata).await? {
-                    // Transition from OutOfSync to InSync state
-                    curr.on_exiting().await;
-
-                    // Enter new state
-                    let mut next = InSyncImpl::new(
-                        self.acc.clone(),
-                        self.network.clone(),
-                        self.blacklisted_blocks.clone(),
-                    );
-                    next.on_entering(blk).await.map_err(|e| {
-                        error!("Unable to enter in_sync state: {e}");
-                        e
-                    })?;
-                    self.curr = State::InSync(next);
+                State::OutOfSync(ref mut curr) => {
+                    if curr.on_block_event(blk, metadata).await? {
+                        // Transition from OutOfSync to InSync state
+                        curr.on_exiting().await;
+
+                        // Enter new state
+                        let mut next = InSyncImpl::new(
+                            self.acc.clone(),
+                            self.network.clone(),
+                            self.blacklisted_blocks.clone(),
+                        );
+                        next.on_entering(blk).await.map_err(|e| {
+                            error!("Unable to enter in_sync state: {e}");
+                            e
+                        })?;
+                        self.curr = State::InSync(next);
+                    }
                 }
             }
         }
+
+        let now = Instant::now();
+        self.certificates_cache
+            .retain(|_, (_, expiry)| *expiry > now);
```

Comment:
`on_idle` is called only in case we don't accept a block for more than 20s.That said, we need to clean up the cert_cache in a regular event like `block_accept`.

Reply:
That's very weird to me. We should have a routine doing this kind of things (cleanup) after each round.
It shouldn't be part of `block_accept`. I guess this requires a separate PR though

Reply:
We have but it does not have access to the `certificate_cache` as it used only in FSM component.

Reply:
Then we should implement something apart (in another PR).
Since we can't do it atm, please add a comment there to explain it.

Reply:
> Then we should implement something apart (in another PR).
Go for it.
It's pretty normal in a msg handler (in Reactor pattern impl)  to do a cleanup of cache/data relevant to the handler/msg.  

> Since we can't do it atm, please add a comment there to explain it.

That comment is only for your benefit, feel free to commit it by yourself.